katello_certs/hooks/boot/02-message-helpers.rb

```diff
@@ -43,7 +45,7 @@ def proxy_instructions_message(kafo)
   If you do not have the #{proxy_name} registered to the #{main_instance_name} instance, then please do the following:
 
   1. yum -y localinstall http://#{fqdn}/pub/katello-ca-consumer-latest.noarch.rpm
-  2. subscription-manager register --org "<%= color('#{org}', :info) %>"
+  2. subscription-manager register --org "<%= color('#{Shellwords.escape(org)}', :info) %>"
```

Comment:
```suggestion
  2. subscription-manager register --org "<%= color(Shellwords.escape(org), :info) %>"
```

Reply:
I tried that originally :/

```
/opt/rh/rh-ruby27/root/usr/share/ruby/erb.rb:905:in `eval': (erb):8: syntax error, unexpected backslash (SyntaxError)
....freeze; _erbout.<<(( color(Te\\"st, :info) ).to_s); _erbout....
...                              ^
(erb):8: syntax error, unexpected backslash, expecting ')'
..., :info) ).to_s); _erbout.<< "\\"\\n".freeze
...                              ^
(erb):28: syntax error, unexpected end-of-input, expecting ')'
```

Reply:
huh, then I need to look at the whole file not from mobile XD

Reply:
It's already pure Ruby since it's within `<%=` and `%>`. So this should work
```suggestion
  2. subscription-manager register --org "<%= color(Shellwords.escape(org), :info) %>"
```

Reply:
This a template with strings inside of it where the template gets passed and evaluated. We need the Shellwords portion evaluated at the time the string itself is declared in the method and not later when passed to the `say` method and hence the interpolation. To show what happens with this suggestion:

```
  2. subscription-manager register --org "Method\ org\ with\ args\ \[\]\ is\ not\ available\ on\ \#\<HighLine::TemplateRenderer:0x000055a1f106e8c8\ @template\=\#\<ERB:0x000055a1f10baae8\ @src\=\"\#coding:UTF-8\\n_erbout\ \=\ +\'\'\;\ _erbout.\<\<\ \\\"\ \ \\\".freeze\;\ _erbout.\<\<\(\(\ color\(\'Success\!\',\ :good\)\ \).to_s\)\;\ _erbout.\<\<\ \\\"\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ To\ finish\ the\ installation,\ follow\ these\ steps:\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ If\ you\ do\ not\ have\ the\ Smart\ Proxy\ registered\ to\ the\ Katello\ instance,\ then\ please\ do\ the\ following:\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ 1.\ yum\ -y\ localinstall\ http://centos8-stream-katello-nightly.wareagle.example.com/pub/katello-ca-consumer-latest.noarch.rpm\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ 2.\ subscription-manager\ register\ --org\ \\\\\\\"\\\".freeze\;\ _erbout.\<\<\(\(\ color\(Shellwords.escape\(org\),\ :info\)\ \).to_s\)\;\ _erbout.\<\<\ \\\"\\\\\\\"\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ Once\ this\ is\ completed\ run\ the\ steps\ below\ to\ start\ the\ Smart\ Proxy\ installation:\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ 1.\ Ensure\ that\ the\ foreman-installer-katello\ package\ is\ installed\ on\ the\ system.\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ 2.\ Copy\ the\ following\ file\ \\\".freeze\;\ _erbout.\<\<\(\(\ color\(\\\"/root/foo.example.com\\\",\ :info\)\ \).to_s\)\;\ _erbout.\<\<\ \\\"\ to\ the\ system\ \\\".freeze\;\ _erbout.\<\<\(\(\ color\(\\\"foo.example.com\\\",\ :info\)\ \).to_s\)\;\ _erbout.\<\<\ \\\"\ at\ the\ following\ location\ \\\".freeze\;\ _erbout.\<\<\(\(\ color\(\\\"/root/foo.example.com\\\",\ :info\)\ \).to_s\)\;\ _erbout.\<\<\ \\\"\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ scp\ \\\".freeze\;\ _erbout.\<\<\(\(\ color\(\\\"/root/foo.example.com\\\",\ :info\)\ \).to_s\)\;\ _erbout.\<\<\ \\\"\ root@\\\".freeze\;\ _erbout.\<\<\(\(\ color\(\\\"foo.example.com\\\",\ :info\)\ \).to_s\)\;\ _erbout.\<\<\ \\\":\\\".freeze\;\ _erbout.\<\<\(\(\ color\(\\\"/root/foo.example.com\\\",\ :info\)\ \).to_s\)\;\ _erbout.\<\<\ \\\"\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ 3.\ Run\ the\ following\ commands\ on\ the\ Smart\ Proxy\ \(possibly\ with\ the\ customized\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ \ \ \ parameters,\ see\ \\\".freeze\;\ _erbout.\<\<\(\(\ color\(\\\"foreman-installer\ --scenario\ foreman-proxy-content\ --help\\\",\ :info\)\ \).to_s\)\;\ _erbout.\<\<\ \\\"\ and\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ \ \ \ documentation\ for\ more\ info\ on\ setting\ up\ additional\ services\):\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ foreman-installer\\\\\\\\\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --scenario\ foreman-proxy-content\\\\\\\\\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --certs-tar-file\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \\\\\\\"\\\".freeze\;\ _erbout.\<\<\(\(\ color\(\\\"/root/foo.example.com\\\",\ :info\)\ \).to_s\)\;\ _erbout.\<\<\ \\\"\\\\\\\"\\\\\\\\\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-register-in-foreman\ \ \ \ \ \ \ \ \ \ \ \\\\\\\"true\\\\\\\"\\\\\\\\\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-foreman-base-url\ \ \ \ \ \ \ \ \ \ \ \ \ \ \\\\\\\"https://centos8-stream-katello-nightly.wareagle.example.com\\\\\\\"\\\\\\\\\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-trusted-hosts\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \\\\\\\"centos8-stream-katello-nightly.wareagle.example.com\\\\\\\"\\\\\\\\\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-trusted-hosts\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \\\\\\\"foo.example.com\\\\\\\"\\\\\\\\\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-oauth-consumer-key\ \ \ \ \ \ \ \ \ \ \ \ \\\\\\\"dJs8tyKxYpyfRS6tpQkcf6B6jtPyemud\\\\\\\"\\\\\\\\\\\\n\\\".freeze\\n\;\ _erbout.\<\<\ \\\"\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-oauth-consumer-secret\ \ \ \ \ \ \ \ \ \\\\\\\"c2PDMHkPNVnAPA855Jk46RatBeFGttTd\\\\\\\"\\\\n\\\".freeze\\n\;\ _erbout\",\ @encoding\=\#\<Encoding:UTF-8\>,\ @frozen_string\=nil,\ @filename\=nil,\ @lineno\=0,\ @_init\=\#\<Class:ERB\>\>,\ @source\=\"\ \ \<\%\=\ color\(\'Success\!\',\ :good\)\ \%\>\\n\\n\ \ To\ finish\ the\ installation,\ follow\ these\ steps:\\n\\n\ \ If\ you\ do\ not\ have\ the\ Smart\ Proxy\ registered\ to\ the\ Katello\ instance,\ then\ please\ do\ the\ following:\\n\\n\ \ 1.\ yum\ -y\ localinstall\ http://centos8-stream-katello-nightly.wareagle.example.com/pub/katello-ca-consumer-latest.noarch.rpm\\n\ \ 2.\ subscription-manager\ register\ --org\ \\\"\<\%\=\ color\(Shellwords.escape\(org\),\ :info\)\ \%\>\\\"\\n\\n\ \ Once\ this\ is\ completed\ run\ the\ steps\ below\ to\ start\ the\ Smart\ Proxy\ installation:\\n\\n\ \ 1.\ Ensure\ that\ the\ foreman-installer-katello\ package\ is\ installed\ on\ the\ system.\\n\ \ 2.\ Copy\ the\ following\ file\ \<\%\=\ color\(\\\"/root/foo.example.com\\\",\ :info\)\ \%\>\ to\ the\ system\ \<\%\=\ color\(\\\"foo.example.com\\\",\ :info\)\ \%\>\ at\ the\ following\ location\ \<\%\=\ color\(\\\"/root/foo.example.com\\\",\ :info\)\ \%\>\\n\ \ scp\ \<\%\=\ color\(\\\"/root/foo.example.com\\\",\ :info\)\ \%\>\ root@\<\%\=\ color\(\\\"foo.example.com\\\",\ :info\)\ \%\>:\<\%\=\ color\(\\\"/root/foo.example.com\\\",\ :info\)\ \%\>\\n\ \ 3.\ Run\ the\ following\ commands\ on\ the\ Smart\ Proxy\ \(possibly\ with\ the\ customized\\n\ \ \ \ \ parameters,\ see\ \<\%\=\ color\(\\\"foreman-installer\ --scenario\ foreman-proxy-content\ --help\\\",\ :info\)\ \%\>\ and\\n\ \ \ \ \ documentation\ for\ more\ info\ on\ setting\ up\ additional\ services\):\\n\\n\ \ foreman-installer\\\\\\n\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --scenario\ foreman-proxy-content\\\\\\n\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --certs-tar-file\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \\\"\<\%\=\ color\(\\\"/root/foo.example.com\\\",\ :info\)\ \%\>\\\"\\\\\\n\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-register-in-foreman\ \ \ \ \ \ \ \ \ \ \ \\\"true\\\"\\\\\\n\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-foreman-base-url\ \ \ \ \ \ \ \ \ \ \ \ \ \ \\\"https://centos8-stream-katello-nightly.wareagle.example.com\\\"\\\\\\n\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-trusted-hosts\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \\\"centos8-stream-katello-nightly.wareagle.example.com\\\"\\\\\\n\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-trusted-hosts\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \\\"foo.example.com\\\"\\\\\\n\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-oauth-consumer-key\ \ \ \ \ \ \ \ \ \ \ \ \\\"dJs8tyKxYpyfRS6tpQkcf6B6jtPyemud\\\"\\\\\\n\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ --foreman-proxy-oauth-consumer-secret\ \ \ \ \ \ \ \ \ \\\"c2PDMHkPNVnAPA855Jk46RatBeFGttTd\\\"\\n\",\ @highline\=\#\<HighLine:0x000055a1f08117c0\ @input\=\#\<IO:\<STDIN\>\>,\ @output\=\#\<IO:\<STDOUT\>\>,\ @multi_indent\=true,\ @indent_size\=3,\ @indent_level\=0,\ @wrap_at\=nil,\ @page_at\=nil,\ @header\=nil,\ @prompt\=nil,\ @key\=nil,\ @use_color\=true,\ @track_eof\=true,\ @terminal\=\#\<HighLine::Terminal::IOConsole:0x000055a1f06122f8\ @input\=\#\<IO:\<STDIN\>\>,\ @output\=\#\<IO:\<STDOUT\>\>\>\>\>.\ Try\ \[\]"
```

Reply:
So I think I have a better understanding of this now. How about this
```suggestion
  2. subscription-manager register --org "<%= #{color(org, :info)} %>"
```

Reply:
Feeling like I am on a merry go round, hoping we can just take my solution as is now :smile: 

The layers can be confusing, so note that `color` is function that exists within the templating tags and within the template rendering that is done by Highline (see examples at https://github.com/JEG2/highline#usage). Whereas the imparting of the organization string into the template is handled within this immediate function within which's context the org variable exists.

```
Traceback (most recent call last):
	15: from /sbin/foreman-proxy-certs-generate:31:in `<main>'
	14: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/kafo_configure.rb:50:in `run'
	13: from /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:132:in `run'
	12: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/kafo_configure.rb:180:in `run'
	11: from /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
	10: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/kafo_configure.rb:216:in `execute'
	 9: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/kafo_configure.rb:545:in `run_installation'
	 8: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/hooking.rb:65:in `execute'
	 7: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/hooking.rb:65:in `each'
	 6: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/hooking.rb:67:in `block in execute'
	 5: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/hook_context.rb:19:in `execute'
	 4: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/hook_context.rb:19:in `instance_eval'
	 3: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/hooking.rb:36:in `block (4 levels) in load'
	 2: from /usr/share/gems/gems/kafo-6.5.0/lib/kafo/hooking.rb:36:in `instance_eval'
	 1: from /usr/share/foreman-installer/katello-certs/hooks/post/10-post_generation.rb:2:in `block (4 levels) in load'
/usr/share/foreman-installer/katello-certs/hooks/boot/02-message-helpers.rb:46:in `proxy_instructions_message': undefined method `color' for #<Kafo::HookContext:0x00005651c21e9cd0> (NoMethodError)
```

Reply:
Absolutely right that this is confusing in the layering. That's what also got me in the first place.

It's also what we had to work around here:
https://github.com/theforeman/foreman-installer/blob/26de96d7589ab37c98a373053216a740523af32b/hooks/boot/01-kafo-hook-extensions.rb#L100

Perhaps we should enhance Kafo to expose the `color` method in hooks but for now I'd use `Highline.color()` directly.

Reply:
Those could both be useful enhancements, given that the usage pattern of `color` inside a template passed to the `say` function is prevalent throughout not only this file but the various hooks directories I would prefer to take this change as is targeted at dealing with just outlined problem of allowing an organization with quotes and what-not to not break this piece of functionality for users.

Reply:
My point is that `Shellwords.escape()` escapes for a shell. You could introduce odd edge cases. A minimal reproducer:

```ruby
require 'erb'

def color(message)
  puts message
end

org = "My Org With ' and ~".tr(' ', '_')
template = "<%= color('#{Shellwords.escape(org)}') %>"
ERB.new(template).result
```

This prints `My_Org_With_'_and_\~` and the reason is this:
```
[1] pry(main)> org = "My Org With ' and ~".tr(' ', '_')
=> "My_Org_With_'_and_~"
[2] pry(main)> template = "<%= color('#{Shellwords.escape(org)}') %>"
=> "<%= color('My_Org_With_\\'_and_\\~') %>"
```

You can see it also escapes `~`.

So I really insist we don't use `Shellwords.escape` for this and avoiding any escaping. I think this should work: skip templating and use regular string interpolation here:
```suggestion
  2. subscription-manager register --org "#{HighLine.color(org, :info)}"
```

Interestingly, this printing unvalidated user input creates a risk of exploits. In theory someone could give you a prepared answers file where the Ruby is escaped and opens a root shell. Granted, very unlikely but you could hide it in the initial admin password.